src/git.rs

```diff
@@ -0,0 +1,268 @@
+//! Git utilities
+
+use std::ffi::OsStr;
+use std::num::NonZeroU64;
+use std::path::{Path, PathBuf};
+
+use anyhow::{anyhow, Result};
+
+use crate::fs::remove_dir_all;
+use crate::{cmd, cmd_output};
+
+pub const GIT: &str = "git";
+
+pub struct Repository {
+    git_dir: PathBuf,
+    worktree: PathBuf,
+    remote_name: Option<String>,
+}
+
+impl Repository {
+    pub fn new(dir: impl AsRef<Path>) -> Repository {
+        Repository {
+            git_dir: dir.as_ref().to_owned().join(".git"),
+            worktree: dir.as_ref().to_owned(),
+            remote_name: None,
+        }
+    }
+
+    /// Get the remote name from which this repository was cloned.
+    pub fn origin(&self) -> Option<&String> {
+        self.remote_name.as_ref()
+    }
+
+    fn git_args(&self) -> [&OsStr; 4] {
```

Comment:
`git apply` does not really work for me (mac) with the `--work-tree` and `--git-dir`.
For example, `cd`-ing to `esp-idf-sys` and typing 
`git --git-dir /Users/imarkov/projects/esp32/rust-esp32-std-hello/.sdk/esp-idf/.git --work-tree /Users/imarkov/projects/esp32/rust-esp32-std-hello/.sdk/esp-idf apply patches/missing_xtensa_atomics_fix.diff` 

returns:
`error: components/xtensa/stdatomic.c: No such file or directory`

If I just do instead:
`cd /Users/imarkov/projects/esp32/rust-esp32-std-hello/.sdk/esp-idf`
`git apply ~/projects/esp32/esp-idf-sys/patches/missing_xtensa_atomics_fix.diff`

It does work.

The above is one of the reason why the native `esp-idf-sys` build fails.

Reply:
That's weird, I've not had this problem on windows (git version `2.27.0`). Maybe the problem is that only the form `--git-dir=<path>` and `--work-tree=<path>` work, I currently pass them as separate arguments (which works in windows). The git docs only mention the `=`-version.

Could you verify if the `=`-version works?

Reply:
Nevermind, I messed up. It never executed on my part. Will fix it.

Reply:
Okay, should be fixed now.